File: go/method_normalization_test.go
package simsysmetrics

import (
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

// TestNormalizeMethod unit-tests the allow-list helper directly.
func TestNormalizeMethod(t *testing.T) {
	cases := []struct {
		input string
		want  string
	}{
		{"GET", "GET"},
		{"get", "GET"},
		{"Post", "POST"},
		{"PATCH", "PATCH"},
		{"OPTIONS", "OPTIONS"},
		{"X_AUDIT_1", "OTHER"},
		{"ASDF", "OTHER"},
		{"", "OTHER"},
		{"BREW", "OTHER"},
		{"PROPFIND", "OTHER"},
	}
	for _, c := range cases {
		if got := NormalizeMethod(c.input); got != c.want {
			t.Errorf("NormalizeMethod(%q) = %q; want %q", c.input, got, c.want)
		}
	}
}

// TestMiddlewareGarbageMethodsCollapseToOTHER is the regression test for
// the v0.3.5 cardinality leak: arbitrary client-supplied HTTP methods
// like "X_AUDIT_1", "ASDF" produced one new label series per distinct
// method value. Now they all collapse to method="OTHER".
func TestMiddlewareGarbageMethodsCollapseToOTHER(t *testing.T) {
	m := mustInstallForTest(t, "method-norm")
	handler := m.Middleware(MiddlewareOpts{
		Extractor: func(*http.Request) string { return "/x" },
	})(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		w.WriteHeader(200)
	}))

	srv := httptest.NewServer(handler)
	defer srv.Close()

	// Send several non-standard methods. The Go client accepts arbitrary
	// HTTP tokens.
	for _, garbage := range []string{"X_AUDIT_1", "ASDF", "BREW"} {
		req, err := http.NewRequest(garbage, srv.URL+"/x", nil)
		if err != nil {
			t.Fatalf("NewRequest %s: %v", garbage, err)
		}
		resp, err := srv.Client().Do(req)
		if err != nil {
			t.Fatalf("Do %s: %v", garbage, err)
		}
		resp.Body.Close()
	}

	body := scrapeMetrics(t, m)

	// Raw garbage method names must NOT appear as label values.
	for _, garbage := range []string{"X_AUDIT_1", "ASDF", "BREW"} {
		needle := `method="` + garbage + `"`
		if strings.Contains(body, needle) {
			t.Errorf("garbage method %q leaked as label value:\n%s", garbage, body)
		}
	}
	// All three must collapse into method="OTHER".
	if !strings.Contains(body, `method="OTHER"`) {
		t.Errorf("expected method=\"OTHER\" bucket; got:\n%s", body)
	}
}

File: node/tests/method-normalization.test.ts
/**
 * HTTP method label must collapse to a bounded allow-list.
 *
 * Pre-fix the raw `req.method` / `c.req.method` was passed straight
 * through, so a hostile client sending arbitrary methods like
 * `X_AUDIT_1`, `ASDF`, etc. would force one new
 * simsys_http_requests_total series per unique method value.
 */

import { describe, it, expect, beforeEach } from "vitest";
import express from "express";
import { Hono } from "hono";
import { install, registry } from "../src/index.js";
import { normalizeMethod } from "../src/registry.js";
import { _resetForTests as _resetProc } from "../src/process.js";
import { _resetForTests as _resetBase } from "../src/baseline.js";

describe("HTTP method normalization", () => {
  beforeEach(() => {
    _resetProc();
    _resetBase();
    registry.resetMetrics();
  });

  it("normalizeMethod unit cases", () => {
    expect(normalizeMethod("GET")).toBe("GET");
    expect(normalizeMethod("get")).toBe("GET");
    expect(normalizeMethod("Post")).toBe("POST");
    expect(normalizeMethod("PATCH")).toBe("PATCH");
    expect(normalizeMethod("OPTIONS")).toBe("OPTIONS");
    expect(normalizeMethod("X_AUDIT_1")).toBe("OTHER");
    expect(normalizeMethod("ASDF")).toBe("OTHER");
    expect(normalizeMethod("")).toBe("OTHER");
    expect(normalizeMethod(undefined)).toBe("OTHER");
    expect(normalizeMethod(null)).toBe("OTHER");
    expect(normalizeMethod(123)).toBe("OTHER");
  });

  it("express: non-allow-listed methods collapse to method=OTHER", async () => {
    const app = express();
    install(app, { service: "express-method-norm", version: "0.0.0" });
    // Catch-all route so Express actually responds to unusual methods —
    // the cardinality fix is applied on the metric path, regardless of
    // whether the route 200s or 404s.
    app.all("/", (_req, res) => res.status(200).send("ok"));

    await new Promise<void>((resolve, reject) => {
      const server = app.listen(0, async () => {
        try {
          const addr = server.address();
          if (!addr || typeof addr === "string") {
            server.close();
            resolve();
            return;
          }
          const base = `http://127.0.0.1:${addr.port}`;

          // BREW (RFC 2324) and MKCALENDAR (CalDAV) are valid HTTP tokens
          // but not in the allow-list. fetch accepts arbitrary tokens.
          for (const garbage of ["BREW", "MKCALENDAR", "PROPFIND"]) {
            await fetch(`${base}/`, { method: garbage });
          }

          const body = await (await fetch(`${base}/metrics`)).text();

          for (const garbage of ["BREW", "MKCALENDAR", "PROPFIND"]) {
            expect(body).not.toContain(`method="${garbage}"`);
          }
          expect(body).toMatch(/method="OTHER"/);

          server.close();
          resolve();
        } catch (err) {
          server.close();
          reject(err);
        }
      });
    });
  });

  it("hono: garbage methods collapse to method=OTHER", async () => {
    const app = new Hono();
    install(app, { service: "hono-method-norm", version: "0.0.0" });
    app.get("/known", (c) => c.text("ok"));

    await app.fetch(new Request("http://l/known"));

    for (const garbage of ["X_AUDIT_1", "ASDF", "UNDEFINED"]) {
      await app.fetch(new Request("http://l/x", { method: garbage }));
    }

    const body = await (await app.fetch(new Request("http://l/metrics"))).text();
    for (const garbage of ["X_AUDIT_1", "ASDF", "UNDEFINED"]) {
      expect(body).not.toContain(`method="${garbage}"`);
    }
    expect(body).toMatch(/method="OTHER"/);
  });
});
